Build error due to minimal libc qsort callback cast

**Describe the bug**
Some toolchains experience problems when casting a function pointer to another function pointer with different number of parameters. This is especially a problem when the cast is from a function pointer with fewer parameters than the function pointer being cast to. As is the case in the minimal libc stdlib.h function qsort.

Following error comes from arm-none-eabi-gcc version 10.3.1 20210824
> stdlib.h:71:29: error: cast between incompatible function types from 'int (*)(const void *, const void )' to 'int ()(const void *, const void *, void *)'

The error above will probably be very similar to what a proprietary toolchain could output.

A fix is proposed in [PR42384](https://github.com/zephyrproject-rtos/zephyr/pull/42384)

**To Reproduce**
The above error was from a zephyr decoupled build of the minimal libc, with the gnu arm compiler. This is only to show that gcc also does not like the cast.

**Expected behavior**
No error during build, even with proprietary toolchains, and thus no casts of this type should be present in the minimal libc.

**Impact**
This could potentially be a showstopper for those not using the zephyr sdk, and thus probably proprietary toolchains.

**Logs and console output**
None

**Environment (please complete the following information):**
 - Linux
 - Toolchain proprietary, but shown with gnu arm gcc version 10.3.1 20210824

**Additional context**
None

--- a/lib/libc/minimal/include/stdlib.h
+++ b/lib/libc/minimal/include/stdlib.h
@@ -32,6 +32,8 @@ void *bsearch(const void *key, const void *array,
 
 void qsort_r(void *base, size_t nmemb, size_t size,
 	     int (*compar)(const void *, const void *, void *), void *arg);
+void qsort(void *base, size_t nmemb, size_t size,
+	   int (*compar)(const void *, const void *));
 
 #define EXIT_SUCCESS 0
 #define EXIT_FAILURE 1
@@ -61,14 +63,6 @@ static inline long labs(long __n)
 static inline long long llabs(long long __n)
 {
 	return (__n < 0LL) ? -__n : __n;
-}
-
-static inline void qsort(void *base, size_t nmemb, size_t size,
-	int (*compar)(const void *, const void *))
-{
-	typedef int (*compar3)(const void *, const void *, void *);
-
-	qsort_r(base, nmemb, size, (compar3)compar, NULL);
 }
 
 #ifdef __cplusplus

--- a/lib/libc/minimal/source/stdlib/qsort.c
+++ b/lib/libc/minimal/source/stdlib/qsort.c
@@ -6,10 +6,9 @@
 
 #include <stddef.h>
 #include <stdint.h>
+#include <stdlib.h>
 #include <sys/types.h>
 #include <sys/util.h>
-
-typedef int (*comp3_t)(const void *, const void *, void *);
 
 /*
  * Normally parent is defined parent(k) = floor((k-1) / 2) but we can avoid a
@@ -32,7 +31,25 @@ typedef int (*comp3_t)(const void *, const void *, void *);
 
 #define A(k) ((uint8_t *)base + size * (k))
 
-static void sift_down(void *base, int start, int end, size_t size, comp3_t comp, void *comp_arg)
+struct qsort_comp {
+	bool has3;
+	void *arg;
+	union {
+		int (*comp2)(const void *a, const void *b);
+		int (*comp3)(const void *a, const void *b, void *arg);
+	};
+};
+
+static inline int compare(struct qsort_comp *cmp, void *a, void *b)
+{
+	if (cmp->has3) {
+		return cmp->comp3(a, b, cmp->arg);
+	}
+
+	return cmp->comp2(a, b);
+}
+
+static void sift_down(void *base, int start, int end, size_t size, struct qsort_comp *cmp)
 {
 	int root;
 	int child;
@@ -42,12 +59,12 @@ static void sift_down(void *base, int start, int end, size_t size, comp3_t comp,
 		child = left(root);
 
 		/* if root < left */
-		if (comp(A(swap), A(child), comp_arg) < 0) {
+		if (compare(cmp, A(swap), A(child)) < 0) {
 			swap = child;
 		}
 
 		/* right exists and min(A(root),A(left)) < A(right) */
-		if (right(root) < end && comp(A(swap), A(right(root)), comp_arg) < 0) {
+		if (right(root) < end && compare(cmp, A(swap), A(right(root))) < 0) {
 			swap = right(root);
 		}
 
@@ -59,28 +76,51 @@ static void sift_down(void *base, int start, int end, size_t size, comp3_t comp,
 	}
 }
 
-static void heapify(void *base, int nmemb, size_t size, comp3_t comp, void *comp_arg)
+static void heapify(void *base, int nmemb, size_t size, struct qsort_comp *cmp)
 {
 	int start;
 
 	for (start = parent(nmemb - 1); start >= 0; --start) {
-		sift_down(base, start, nmemb, size, comp, comp_arg);
+		sift_down(base, start, nmemb, size, cmp);
 	}
 }
 
-static void heap_sort(void *base, int nmemb, size_t size, comp3_t comp, void *comp_arg)
+static void heap_sort(void *base, int nmemb, size_t size, struct qsort_comp *cmp)
 {
 	int end;
 
-	heapify(base, nmemb, size, comp, comp_arg);
+	heapify(base, nmemb, size, cmp);
 
 	for (end = nmemb - 1; end > 0; --end) {
 		byteswp(A(end), A(0), size);
-		sift_down(base, 0, end, size, comp, comp_arg);
+		sift_down(base, 0, end, size, cmp);
 	}
 }
 
-void qsort_r(void *base, size_t nmemb, size_t size, comp3_t comp, void *arg)
+void qsort_r(void *base, size_t nmemb, size_t size,
+	     int (*comp3)(const void *a, const void *b, void *arg), void *arg)
 {
-	heap_sort(base, nmemb, size, comp, arg);
+	struct qsort_comp cmp = {
+		.has3 = true,
+		.arg = arg,
+		{
+			.comp3 = comp3
+		}
+	};
+
+	heap_sort(base, nmemb, size, &cmp);
+}
+
+void qsort(void *base, size_t nmemb, size_t size,
+	   int (*comp2)(const void *a, const void *b))
+{
+	struct qsort_comp cmp = {
+		.has3 = false,
+		.arg = NULL,
+		{
+			.comp2 = comp2
+		}
+	};
+
+	heap_sort(base, nmemb, size, &cmp);
 }

